// angular2/src/test_lib/test_lib.js
"use strict";
Object.defineProperties(module.exports, {
  proxy: {get: function() {
      return $__rtts_95_assert_47_rtts_95_assert__.proxy;
    }},
  describe: {get: function() {
      return describe;
    }},
  xdescribe: {get: function() {
      return xdescribe;
    }},
  ddescribe: {get: function() {
      return ddescribe;
    }},
  it: {get: function() {
      return it;
    }},
  xit: {get: function() {
      return xit;
    }},
  iit: {get: function() {
      return iit;
    }},
  beforeEach: {get: function() {
      return beforeEach;
    }},
  afterEach: {get: function() {
      return afterEach;
    }},
  expect: {get: function() {
      return expect;
    }},
  IS_DARTIUM: {get: function() {
      return IS_DARTIUM;
    }},
  SpyObject: {get: function() {
      return SpyObject;
    }},
  __esModule: {value: true}
});
var __moduleName = "angular2/src/test_lib/test_lib";
var $__angular2_47_src_47_facade_47_dom__,
    $__rtts_95_assert_47_rtts_95_assert__;
var DOM = ($__angular2_47_src_47_facade_47_dom__ = require("angular2/src/facade/dom"), $__angular2_47_src_47_facade_47_dom__ && $__angular2_47_src_47_facade_47_dom__.__esModule && $__angular2_47_src_47_facade_47_dom__ || {default: $__angular2_47_src_47_facade_47_dom__}).DOM;
var $__rtts_95_assert_47_rtts_95_assert__ = ($__rtts_95_assert_47_rtts_95_assert__ = require("rtts_assert/rtts_assert"), $__rtts_95_assert_47_rtts_95_assert__ && $__rtts_95_assert_47_rtts_95_assert__.__esModule && $__rtts_95_assert_47_rtts_95_assert__ || {default: $__rtts_95_assert_47_rtts_95_assert__});
var describe = window.describe;
var xdescribe = window.xdescribe;
var ddescribe = window.ddescribe;
var it = window.it;
var xit = window.xit;
var iit = window.iit;
var beforeEach = window.beforeEach;
var afterEach = window.afterEach;
var expect = window.expect;
var IS_DARTIUM = false;
window.print = function(msg) {
  if (window.dump) {
    window.dump(msg);
  } else {
    window.console.log(msg);
  }
};
window.Map.prototype.jasmineToString = function() {
  var m = this;
  if (!m) {
    return '' + m;
  }
  var res = [];
  m.forEach((function(v, k) {
    res.push((k + ":" + v));
  }));
  return ("{ " + res.join(',') + " }");
};
window.beforeEach(function() {
  jasmine.addMatchers({
    toEqual: function(util, customEqualityTesters) {
      return {compare: function(actual, expected) {
          return {pass: util.equals(actual, expected, [compareMap])};
        }};
      function compareMap(actual, expected) {
        if (actual instanceof Map) {
          var pass = actual.size === expected.size;
          if (pass) {
            actual.forEach((function(v, k) {
              pass = pass && util.equals(v, expected.get(k));
            }));
          }
          return pass;
        } else {
          return undefined;
        }
      }
    },
    toBePromise: function() {
      return {compare: function(actual, expectedClass) {
          var pass = typeof actual === 'object' && typeof actual.then === 'function';
          return {
            pass: pass,
            get message() {
              return 'Expected ' + actual + ' to be a promise';
            }
          };
        }};
    },
    toBeAnInstanceOf: function() {
      return {compare: function(actual, expectedClass) {
          var pass = typeof actual === 'object' && actual instanceof expectedClass;
          return {
            pass: pass,
            get message() {
              return 'Expected ' + actual + ' to be an instance of ' + expectedClass;
            }
          };
        }};
    },
    toHaveText: function() {
      return {compare: function(actual, expectedText) {
          var actualText = elementText(actual);
          return {
            pass: actualText == expectedText,
            get message() {
              return 'Expected ' + actualText + ' to be equal to ' + expectedText;
            }
          };
        }};
    },
    toImplement: function() {
      return {compare: function(actualObject, expectedInterface) {
          var objProps = Object.keys(actualObject.constructor.prototype);
          var intProps = Object.keys(expectedInterface.prototype);
          var missedMethods = [];
          intProps.forEach((function(k) {
            if (!actualObject.constructor.prototype[k])
              missedMethods.push(k);
          }));
          return {
            pass: missedMethods.length == 0,
            get message() {
              return 'Expected ' + actualObject + ' to have the following methods: ' + missedMethods.join(", ");
            }
          };
        }};
    }
  });
});
var SpyObject = function SpyObject() {};
($traceurRuntime.createClass)(SpyObject, {
  spy: function(name) {
    if (!this[name]) {
      this[name] = this._createGuinnessCompatibleSpy();
    }
    return this[name];
  },
  rttsAssert: function(value) {
    return true;
  },
  _createGuinnessCompatibleSpy: function() {
    var newSpy = jasmine.createSpy();
    newSpy.andCallFake = newSpy.and.callFake;
    return newSpy;
  }
}, {});
function elementText(n) {
  var hasShadowRoot = (function(n) {
    return n instanceof Element && n.shadowRoot;
  });
  var hasNodes = (function(n) {
    var children = DOM.childNodes(n);
    return children && children.length > 0;
  });
  if (n instanceof Comment)
    return '';
  if (n instanceof Array)
    return n.map((function(nn) {
      return elementText(nn);
    })).join("");
  if (n instanceof Element && DOM.tagName(n) == 'CONTENT')
    return elementText(Array.prototype.slice.apply(n.getDistributedNodes()));
  if (hasShadowRoot(n))
    return elementText(DOM.childNodesAsList(n.shadowRoot));
  if (hasNodes(n))
    return elementText(DOM.childNodesAsList(n));
  return n.textContent;
}
