// angular2/src/core/compiler/shadow_dom_emulation/content_tag.js
"use strict";
Object.defineProperties(module.exports, {
  Content: {get: function() {
      return Content;
    }},
  __esModule: {value: true}
});
var __moduleName = "angular2/src/core/compiler/shadow_dom_emulation/content_tag";
var $__rtts_95_assert_47_rtts_95_assert__,
    $__angular2_47_src_47_core_47_annotations_47_annotations__,
    $__angular2_47_src_47_core_47_compiler_47_shadow_95_dom_95_emulation_47_light_95_dom__,
    $__angular2_47_di__,
    $__angular2_47_src_47_facade_47_dom__,
    $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_facade_47_collection__,
    $__angular2_47_src_47_core_47_dom_47_element__;
var assert = ($__rtts_95_assert_47_rtts_95_assert__ = require("rtts_assert/rtts_assert"), $__rtts_95_assert_47_rtts_95_assert__ && $__rtts_95_assert_47_rtts_95_assert__.__esModule && $__rtts_95_assert_47_rtts_95_assert__ || {default: $__rtts_95_assert_47_rtts_95_assert__}).assert;
var Decorator = ($__angular2_47_src_47_core_47_annotations_47_annotations__ = require("../../annotations/annotations"), $__angular2_47_src_47_core_47_annotations_47_annotations__ && $__angular2_47_src_47_core_47_annotations_47_annotations__.__esModule && $__angular2_47_src_47_core_47_annotations_47_annotations__ || {default: $__angular2_47_src_47_core_47_annotations_47_annotations__}).Decorator;
var $__2 = ($__angular2_47_src_47_core_47_compiler_47_shadow_95_dom_95_emulation_47_light_95_dom__ = require("./light_dom"), $__angular2_47_src_47_core_47_compiler_47_shadow_95_dom_95_emulation_47_light_95_dom__ && $__angular2_47_src_47_core_47_compiler_47_shadow_95_dom_95_emulation_47_light_95_dom__.__esModule && $__angular2_47_src_47_core_47_compiler_47_shadow_95_dom_95_emulation_47_light_95_dom__ || {default: $__angular2_47_src_47_core_47_compiler_47_shadow_95_dom_95_emulation_47_light_95_dom__}),
    SourceLightDom = $__2.SourceLightDom,
    DestinationLightDom = $__2.DestinationLightDom,
    LightDom = $__2.LightDom;
var Inject = ($__angular2_47_di__ = require("angular2/di"), $__angular2_47_di__ && $__angular2_47_di__.__esModule && $__angular2_47_di__ || {default: $__angular2_47_di__}).Inject;
var $__4 = ($__angular2_47_src_47_facade_47_dom__ = require("angular2/src/facade/dom"), $__angular2_47_src_47_facade_47_dom__ && $__angular2_47_src_47_facade_47_dom__.__esModule && $__angular2_47_src_47_facade_47_dom__ || {default: $__angular2_47_src_47_facade_47_dom__}),
    Element = $__4.Element,
    Node = $__4.Node,
    DOM = $__4.DOM;
var isPresent = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}).isPresent;
var $__6 = ($__angular2_47_src_47_facade_47_collection__ = require("angular2/src/facade/collection"), $__angular2_47_src_47_facade_47_collection__ && $__angular2_47_src_47_facade_47_collection__.__esModule && $__angular2_47_src_47_facade_47_collection__ || {default: $__angular2_47_src_47_facade_47_collection__}),
    List = $__6.List,
    ListWrapper = $__6.ListWrapper;
var NgElement = ($__angular2_47_src_47_core_47_dom_47_element__ = require("angular2/src/core/dom/element"), $__angular2_47_src_47_core_47_dom_47_element__ && $__angular2_47_src_47_core_47_dom_47_element__.__esModule && $__angular2_47_src_47_core_47_dom_47_element__ || {default: $__angular2_47_src_47_core_47_dom_47_element__}).NgElement;
var _scriptTemplate = DOM.createScriptTag('type', 'ng/content');
var ContentStrategy = function ContentStrategy() {};
($traceurRuntime.createClass)(ContentStrategy, {insert: function(nodes) {
    assert.argumentTypes(nodes, assert.genericType(List, Node));
  }}, {});
Object.defineProperty(ContentStrategy.prototype.insert, "parameters", {get: function() {
    return [[assert.genericType(List, Node)]];
  }});
var RenderedContent = function RenderedContent(contentEl) {
  assert.argumentTypes(contentEl, Element);
  $traceurRuntime.superConstructor($RenderedContent).call(this);
  this._replaceContentElementWithScriptTags(contentEl);
  this.nodes = [];
};
var $RenderedContent = RenderedContent;
($traceurRuntime.createClass)(RenderedContent, {
  insert: function(nodes) {
    assert.argumentTypes(nodes, assert.genericType(List, Node));
    this.nodes = nodes;
    DOM.insertAllBefore(this.endScript, nodes);
    this._removeNodesUntil(ListWrapper.isEmpty(nodes) ? this.endScript : nodes[0]);
  },
  _replaceContentElementWithScriptTags: function(contentEl) {
    assert.argumentTypes(contentEl, Element);
    this.beginScript = DOM.clone(_scriptTemplate);
    this.endScript = DOM.clone(_scriptTemplate);
    DOM.insertBefore(contentEl, this.beginScript);
    DOM.insertBefore(contentEl, this.endScript);
    DOM.removeChild(DOM.parentElement(contentEl), contentEl);
  },
  _removeNodesUntil: function(node) {
    assert.argumentTypes(node, Node);
    var p = DOM.parentElement(this.beginScript);
    for (var next = DOM.nextSibling(this.beginScript); next !== node; next = DOM.nextSibling(this.beginScript)) {
      DOM.removeChild(p, next);
    }
  }
}, {}, ContentStrategy);
Object.defineProperty(RenderedContent, "parameters", {get: function() {
    return [[Element]];
  }});
Object.defineProperty(RenderedContent.prototype.insert, "parameters", {get: function() {
    return [[assert.genericType(List, Node)]];
  }});
Object.defineProperty(RenderedContent.prototype._replaceContentElementWithScriptTags, "parameters", {get: function() {
    return [[Element]];
  }});
Object.defineProperty(RenderedContent.prototype._removeNodesUntil, "parameters", {get: function() {
    return [[Node]];
  }});
var IntermediateContent = function IntermediateContent(destinationLightDom) {
  assert.argumentTypes(destinationLightDom, LightDom);
  $traceurRuntime.superConstructor($IntermediateContent).call(this);
  this.destinationLightDom = destinationLightDom;
  this.nodes = [];
};
var $IntermediateContent = IntermediateContent;
($traceurRuntime.createClass)(IntermediateContent, {insert: function(nodes) {
    assert.argumentTypes(nodes, assert.genericType(List, Node));
    this.nodes = nodes;
    this.destinationLightDom.redistribute();
  }}, {}, ContentStrategy);
Object.defineProperty(IntermediateContent, "parameters", {get: function() {
    return [[LightDom]];
  }});
Object.defineProperty(IntermediateContent.prototype.insert, "parameters", {get: function() {
    return [[assert.genericType(List, Node)]];
  }});
var Content = function Content(destinationLightDom, contentEl) {
  assert.argumentTypes(destinationLightDom, assert.type.any, contentEl, NgElement);
  this.select = contentEl.getAttribute('select');
  this._strategy = isPresent(destinationLightDom) ? new IntermediateContent(destinationLightDom) : new RenderedContent(contentEl.domElement);
};
($traceurRuntime.createClass)(Content, {
  nodes: function() {
    return assert.returnType((this._strategy.nodes), assert.genericType(List, Node));
  },
  insert: function(nodes) {
    assert.argumentTypes(nodes, assert.genericType(List, Node));
    this._strategy.insert(nodes);
  }
}, {});
Object.defineProperty(Content, "annotations", {get: function() {
    return [new Decorator({selector: 'content'})];
  }});
Object.defineProperty(Content, "parameters", {get: function() {
    return [[new Inject(DestinationLightDom)], [NgElement]];
  }});
Object.defineProperty(Content.prototype.insert, "parameters", {get: function() {
    return [[assert.genericType(List, Node)]];
  }});
